base/main_test_bucket_pool.go

```diff
@@ -0,0 +1,650 @@
+			atomic.AddInt32(&tbp.stats.TotalBucketReadierCount, 1)
+			ctx := bucketNameCtx(ctx, string(testBucketName))
+			tbp.Logf(ctx, "bucketReadier got bucket")
+
+			go func(testBucketName tbpBucketName) {
+				// We might not actually be "done" with the bucket if something fails,
+				// but we need to release the waitgroup so tbp.Close() doesn't block forever.
+				defer tbp.bucketReadierWaitGroup.Done()
+
+				start := time.Now()
+				b, err := tbp.openTestBucket(testBucketName, CreateSleeperFunc(5, 1000))
+				if err != nil {
+					tbp.Logf(ctx, "Couldn't open bucket to get ready, got error: %v", err)
+					return
+				}
+
+				err, _ = RetryLoop(b.GetName()+"bucketReadierRetry", func() (bool, error, interface{}) {
+					tbp.Logf(ctx, "Running bucket through readier function")
+					err = bucketReadierFunc(ctx, b, tbp)
+					if err != nil {
+						tbp.Logf(ctx, "Couldn't ready bucket, got error: %v - Retrying", err)
+						return true, err, nil
+					}
+					return false, nil, nil
+				}, CreateSleeperFunc(5, 1000))
+				if err != nil {
+					tbp.Logf(ctx, "Couldn't ready bucket, got error: %v - Aborting readier for bucket", err)
+					return
+				}
+
+				tbp.Logf(ctx, "Bucket ready, putting back into ready pool")
+				tbp.readyBucketPool <- b
+				atomic.AddInt64(&tbp.stats.TotalBucketReadierDurationNano, time.Since(start).Nanoseconds())
+			}(testBucketName)
+		}
+	}
+
+	tbp.Logf(context.Background(), "Stopped bucketReadier")
+}
+
+// openTestBucket opens the bucket of the given name for the gocb cluster in the given GocbTestBucketPool.
+func (tbp *GocbTestBucketPool) openTestBucket(testBucketName tbpBucketName, sleeper RetrySleeper) (*CouchbaseBucketGoCB, error) {
+
+	ctx := bucketNameCtx(context.Background(), string(testBucketName))
+
+	bucketSpec := tbp.defaultBucketSpec
+	bucketSpec.BucketName = string(testBucketName)
+
+	waitForNewBucketWorker := func() (shouldRetry bool, err error, value interface{}) {
+		gocbBucket, err := GetCouchbaseBucketGoCBFromAuthenticatedCluster(tbp.cluster, bucketSpec, "")
+		if err != nil {
+			tbp.Logf(ctx, "Retrying OpenBucket")
+			return true, err, nil
+		}
+		return false, nil, gocbBucket
+	}
+
+	tbp.Logf(ctx, "Opening bucket")
+	err, val := RetryLoop("waitForNewBucket", waitForNewBucketWorker, sleeper)
+
+	gocbBucket, _ := val.(*CouchbaseBucketGoCB)
+
+	return gocbBucket, err
+}
+
+// TBPBucketInitFunc is a function that is run once (synchronously) when creating/opening a bucket.
+type TBPBucketInitFunc func(ctx context.Context, b Bucket, tbp *GocbTestBucketPool) error
+
+// NoopInitFunc does nothing to init a bucket. This can be used in conjunction with FlushBucketReadier when there's no requirement for views/GSI.
+var NoopInitFunc TBPBucketInitFunc = func(ctx context.Context, b Bucket, tbp *GocbTestBucketPool) error {
+	return nil
+}
+
+// PrimaryIndexInitFunc creates a primary index on the given bucket. This can then be used with N1QLBucketEmptierFunc, for improved compatibility with GSI.
```

Comment:
It doesn't look like this is being used any more, as primary index creation is in InitializeIndexes.

Reply:
This and `N1QLBucketEmptierFunc` will be required for non-GSI packages (`auth` and `base`) when we re-enable GSI and want to empty buckets without using flush.